odl/space/fspace.py

```diff
@@ -25,51 +25,254 @@
 
 # External imports
 import numpy as np
+from functools import wraps
+from inspect import isclass, isfunction
+from itertools import product
 
 # ODL imports
-from odl.operator.operator import Operator
-from odl.set.domain import IntervalProd
-from odl.set.sets import RealNumbers, ComplexNumbers, Set
+from odl.operator.operator import Operator, _dispatch_call_args
+from odl.set.sets import RealNumbers, ComplexNumbers, Set, Field
 from odl.set.space import LinearSpace, LinearSpaceVector
+from odl.util.utility import (is_valid_input_array, is_valid_input_meshgrid,
+                              meshgrid_input_order, vecs_from_meshgrid)
 
 
 __all__ = ('FunctionSet', 'FunctionSetVector',
            'FunctionSpace', 'FunctionSpaceVector')
 
 
+def _out_of_place_not_impl(x, **kwargs):
+    """Dummy function used when out-of-place function is not given."""
+    raise NotImplementedError('no out-of-place evaluation method defined.')
+
+
+def _default_in_place(func, x, out, **kwargs):
+    """Default in-place evaluation method."""
+    out[:] = func(x, **kwargs)
+    return out
+
+
+def _default_out_of_place(func, dtype, x, **kwargs):
+    """Default in-place evaluation method."""
+    if is_valid_input_array(x, func.domain.ndim):
+        if func.domain.ndim == 1:
+            if x.ndim == 2:
+                x = x[0]
+            out_shape = x.shape
+        else:
+            out_shape = (x.shape[1],)
+    elif is_valid_input_meshgrid(x, func.domain.ndim):
+        # Broadcasting fails for only one vector (ndim == 1)
+        if func.domain.ndim == 1:
+            x = x[0]
+            out_shape = x.shape
+        else:
+            out_shape = np.broadcast(*x).shape
+    else:
+        raise TypeError('cannot use in-place method to implement '
+                        'out-of-place non-vectorized evaluation.')
+
+    out = np.empty(out_shape, dtype=dtype)
+    func(x, out=out, **kwargs)
+    return out
+
+
+def vectorize(dtype, outarg='none'):
```

Comment:
Looks very much like something for utility? Perhaps even move this and the `is_valid_...` things to a new file there?